android/src/main/java/com/learnium/RNDeviceInfo/RNDeviceModule.java

```diff
@@ -144,31 +236,294 @@ private String getCurrentCountry() {
+    TelephonyManager telephonyManager =
+            (TelephonyManager) reactContext.getSystemService(Context.TELEPHONY_SERVICE);
+    if (ActivityCompat.checkSelfPermission(reactContext, Manifest.permission.READ_PHONE_STATE) != PackageManager.PERMISSION_GRANTED) {
+      @SuppressLint("HardwareIds") String imsi = telephonyManager.getSubscriberId();
+
+      for (String known_imsi : IMSI_IDS) {
+        if (known_imsi.equalsIgnoreCase(imsi)) {
+          return true;
+        }
+      }
+    }
+
+    return false;
+  }
+
+  private boolean checkOperatorNameAndroid() {
+    String operatorName = ((TelephonyManager)
+            reactContext.getSystemService(Context.TELEPHONY_SERVICE)).getNetworkOperatorName();
+    if (operatorName.equalsIgnoreCase("android")) {
+
+      return true;
+    }
+    return false;
+  }
+
+  private boolean checkDeviceId() {
+    TelephonyManager telephonyManager =
+            (TelephonyManager) reactContext.getSystemService(Context.TELEPHONY_SERVICE);
+
+    if (ActivityCompat.checkSelfPermission(reactContext, Manifest.permission.READ_PHONE_STATE) != PackageManager.PERMISSION_GRANTED) {
+      @SuppressLint("HardwareIds") String deviceId = telephonyManager.getDeviceId();
+
+      for (String known_deviceId : DEVICE_IDS) {
+        if (known_deviceId.equalsIgnoreCase(deviceId)) {
+          return true;
+        }
 
+      }
+    }
+
+    return false;
+  }
+
+  private boolean checkTelephony() {
+    if (ContextCompat.checkSelfPermission(reactContext, Manifest.permission.READ_PHONE_STATE)
+            == PackageManager.PERMISSION_GRANTED && this.isTelephony && isSupportTelePhony()) {
+      return checkPhoneNumber()
+              || checkDeviceId()
+              || checkImsi()
+              || checkOperatorNameAndroid();
+    }
+    return false;
+  }
+
+  private boolean checkQEmuDrivers() {
+    for (File drivers_file : new File[]{new File("/proc/tty/drivers"), new File("/proc/cpuinfo")}) {
+      if (drivers_file.exists() && drivers_file.canRead()) {
+        byte[] data = new byte[1024];
+        try {
+          InputStream is = new FileInputStream(drivers_file);
+          is.read(data);
+          is.close();
+        } catch (Exception exception) {
+          exception.printStackTrace();
+        }
+
+        String driver_data = new String(data);
+        for (String known_qemu_driver : QEMU_DRIVERS) {
+          if (driver_data.contains(known_qemu_driver)) {
+            return true;
+          }
+        }
+      }
+    }
+
+    return false;
+  }
+
+  private boolean checkFiles(String[] targets, String type) {
+    for (String pipe : targets) {
+      File qemu_file = new File(pipe);
+      if (qemu_file.exists()) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  private String getProp(Context context, String property) {
+    try {
+      ClassLoader classLoader = context.getClassLoader();
+      Class<?> systemProperties = classLoader.loadClass("android.os.SystemProperties");
```

Comment:
I didn't think we had to classload anything to get system properties? I thought they were available already?

Reply:
No, to check properties that I declared it needs to load system properties, this class is hide from source. Theres two ways to use this class one by reflection and another by removing _@hide_ using custom SDK. Although **getProp** method it's been used at the **checkQEmuProps** that I wondering I we should keep it or get rid off.

[https://android.googlesource.com/platform/frameworks/base/+/eclair-release/core/java/android/os/SystemProperties.java](url)

Reply:
Ok, so I just researched this, it's in the greylist. I'm not going to add any change like this. https://developer.android.com/about/versions/10/non-sdk-q

You should know any time you need reflection that you're doing something iffy ;-)

Reply:
Okay, as I said I'll remove this.